hedera-node/src/main/java/com/hedera/services/state/expiry/removal/ContractGC.java

```diff
@@ -16,72 +16,85 @@
 
-    // Revisit after release 0.27 to see if VirtualMap.getForModify() has been rehabilitated
     private ContractGC.RemovalFacilitation removalFacilitation =
             MapValueListUtils::removeFromMapValueList;
 
     @Inject
     public ContractGC(
-            final GlobalDynamicProperties dynamicProperties,
+            final ExpiryThrottle expiryThrottle,
             final Supplier<MerkleMap<EntityNum, MerkleAccount>> contracts,
             final Supplier<VirtualMap<ContractKey, IterableContractValue>> storage,
             final Supplier<VirtualMap<VirtualBlobKey, VirtualBlobValue>> bytecode) {
+        this.expiryThrottle = expiryThrottle;
         this.contracts = contracts;
         this.storage = storage;
         this.bytecode = bytecode;
-        this.dynamicProperties = dynamicProperties;
     }
 
     public boolean expireBestEffort(
             final EntityNum expiredContractNum, final MerkleAccount contract) {
-        removeBytecodeIfPresent(expiredContractNum);
         final var numKvPairs = contract.getNumContractKvPairs();
+        boolean rootUpdateThrottled = false;
         if (numKvPairs > 0) {
-            final var maxPairs =
-                    Math.min(numKvPairs, dynamicProperties.getMaxPurgedKvPairsPerTouch());
-            final var removalMeta =
-                    removeKvPairs(
-                            maxPairs,
-                            expiredContractNum,
-                            contract.getFirstContractStorageKey(),
-                            storage.get());
-            final var numRemoved = removalMeta.getKey();
-            if (numRemoved < numKvPairs) {
-                final var mutableContract = contracts.get().getForModify(expiredContractNum);
-                mutableContract.setNumContractKvPairs(numKvPairs - numRemoved);
-                mutableContract.setFirstUint256StorageKey(removalMeta.getValue().getKey());
-                return false;
+            rootUpdateThrottled = !expiryThrottle.allow(ROOT_KEY_UPDATE_WORK);
+            if (!rootUpdateThrottled) {
+                final var slotRemovals =
+                        removeKvPairs(
+                                numKvPairs,
+                                expiredContractNum,
+                                contract.getFirstContractStorageKey(),
+                                storage.get());
+                final var numRemoved = slotRemovals.numRemoved();
+                if (numRemoved < numKvPairs) {
+                    if (numRemoved == 0) {
+                        expiryThrottle.reclaimLastAllowedUse();
+                    } else {
+                        final var mutableContract =
+                                contracts.get().getForModify(expiredContractNum);
+                        mutableContract.setNumContractKvPairs(numKvPairs - numRemoved);
+                        mutableContract.setFirstUint256StorageKey(slotRemovals.newRoot().getKey());
```

Comment:
Codecov shows this condition is not covered

Reply:
Covered now.